### PR TITLE
fix(block_hash): Block hash mismatch for transaction with an empty signature

### LINKED ISSUE #326: Block hash mismatch error while syncing Madara.
### Is there an existing issue?

- [X] I have searched the existing issues

### Description of bug

After running madara as full node, there is an error while syncing:
```
[2024-10-11 11:49:48 INFO] :sparkles: Imported #218483 (0x146c0b...531539) and updated state root (0x38a6a6...143729)
[2024-10-11 11:49:48 INFO] :hourglass_flowing_sand: Gracefully closing the database...
Error: Block hash mismatch: expected 0x495aaa2ae4e2784fcdf4c50a5d21415e388c593a3d0b6db0958758dadd829d9, got 0x7646fc987f5ecc3941804c3fabc6bf684691756d5b2fd21f94fc0b114459188
```

### Steps to reproduce

Run madara with these flags
```
- "--name"
        - "madara"
        - "--base-path"
        - "/madara/data"
        - "--network"
        - "test"
        - "--l1-endpoint"
        - "http://anvil-service:8545"
        - "--full"
        - "--gateway-key"
        - "$(GATEWAY_KEY)"
        ```

### PR DESCRIPTION
**Pull request by Trantorian1**
# Pull Request type

- Bugfix

## What is the current behavior?

Resolves: #326

> [!WARNING]
> Madara was incorrectly calculating the transaction hash with signature for transactions >= v0.13.2 with empty signature.

## What is the new behavior?

Updated and refactored transaction hash with signature computation to take this into account.

## Does this introduce a breaking change?

No.

## Other information

Leaving [this file from Pathfinder](https://github.com/eqlabs/pathfinder/blob/main/crates/pathfinder/src/state/block_hash.rs) for future reference on computing the block hash.

### CODE REVIEW

File: crates/primitives/transactions/src/compute_hash.rs [hunk 1]
```diff
@@ -58,60 +58,53 @@ impl Transaction {
...
+        }
+    }
+
+    fn compute_hash_with_signature_pre_v0_11_1(&self, tx_hash: Felt) -> Felt {
+        let signature_hash = match self {
+            Transaction::Invoke(tx) => tx.compute_hash_signature::<Pedersen>(),
+            Transaction::Declare(_)
+            | Transaction::DeployAccount(_)
+            | Transaction::Deploy(_)
+            | Transaction::L1Handler(_) => Pedersen::hash_array(&[]),
```
Comment (jbcaron) [inline crates/primitives/transactions/src/compute_hash.rs]: create a constant with this value `Pedersen::hash_array(&[])`

Comment (Trantorian1) [reply crates/primitives/transactions/src/compute_hash.rs]: Resolved.

**Review by jbcaron — APPROVED** [review]

**Review by akhercha — APPROVED** [review]
Looks good - nothing to add, thanks for the changes! 🙏

**Comment by Mohiiit** [comment]
we ran into this when we were trying to sync pathfinder with madara as sequencer, and we thought it was wrong on the pathfinder end lol

thanks for this <3

**Pull request merged**